engine/src/flutter/shell/platform/windows/flutter_host_window.cc

```diff
@@ -0,0 +1,519 @@
+         0;
+}
+
+// Converts std::string to std::wstring.
+std::wstring StringToWstring(std::string_view str) {
+  if (str.empty()) {
+    return {};
+  }
+  if (int buffer_size =
+          MultiByteToWideChar(CP_UTF8, 0, str.data(), str.size(), nullptr, 0)) {
+    std::wstring wide_str(buffer_size, L'\0');
+    if (MultiByteToWideChar(CP_UTF8, 0, str.data(), str.size(), &wide_str[0],
+                            buffer_size)) {
+      return wide_str;
+    }
+  }
+  return {};
+}
+
+// Window attribute that enables dark mode window decorations.
+//
+// Redefined in case the developer's machine has a Windows SDK older than
+// version 10.0.22000.0.
+// See:
+// https://docs.microsoft.com/windows/win32/api/dwmapi/ne-dwmapi-dwmwindowattribute
+#ifndef DWMWA_USE_IMMERSIVE_DARK_MODE
+#define DWMWA_USE_IMMERSIVE_DARK_MODE 20
+#endif
+
+// Updates the window frame's theme to match the system theme.
+void UpdateTheme(HWND window) {
+  // Registry key for app theme preference.
+  const wchar_t kGetPreferredBrightnessRegKey[] =
+      L"Software\\Microsoft\\Windows\\CurrentVersion\\Themes\\Personalize";
+  const wchar_t kGetPreferredBrightnessRegValue[] = L"AppsUseLightTheme";
+
+  // A value of 0 indicates apps should use dark mode. A non-zero or missing
+  // value indicates apps should use light mode.
+  DWORD light_mode;
+  DWORD light_mode_size = sizeof(light_mode);
+  LSTATUS const result =
+      RegGetValue(HKEY_CURRENT_USER, kGetPreferredBrightnessRegKey,
+                  kGetPreferredBrightnessRegValue, RRF_RT_REG_DWORD, nullptr,
+                  &light_mode, &light_mode_size);
+
+  if (result == ERROR_SUCCESS) {
+    BOOL enable_dark_mode = light_mode == 0;
+    DwmSetWindowAttribute(window, DWMWA_USE_IMMERSIVE_DARK_MODE,
+                          &enable_dark_mode, sizeof(enable_dark_mode));
+  }
+}
+
+// Inserts |content| into the window tree.
+void SetChildContent(HWND content, HWND window) {
+  SetParent(content, window);
+  RECT client_rect;
+  GetClientRect(window, &client_rect);
+  MoveWindow(content, client_rect.left, client_rect.top,
+             client_rect.right - client_rect.left,
+             client_rect.bottom - client_rect.top, true);
+}
+
+}  // namespace
+
+namespace flutter {
+
+std::unique_ptr<FlutterHostWindow> FlutterHostWindow::createRegularWindow(
+    WindowManager* window_manager,
+    FlutterWindowsEngine* engine,
+    const FlutterWindowSizing& content_size) {
+  DWORD window_style = WS_OVERLAPPEDWINDOW;
+  DWORD extended_window_style = 0;
+  std::optional<Size> smallest = std::nullopt;
+  std::optional<Size> biggest = std::nullopt;
+
+  if (content_size.has_constraints) {
+    smallest = Size(content_size.min_width, content_size.min_height);
+    if (content_size.max_width > 0 && content_size.max_height > 0) {
+      biggest = Size(content_size.max_width, content_size.max_height);
+    }
+  }
+
+  // TODO(knopp): What about windows sized to content?
```

Comment:
I think sizing these windows will be a multi-step process. The frame pipeline will do something like:

1. The framework creates a window with some random size and loose constraints.
2. The embedder creates a hosted window and a view
3. The framework inserts the window's view into the widget tree.
4. The framework lays out the view's content using the loose constraints it chose above
5. The framework calls `FlutterView.render` with a `Scene` whose size fits within the view's loose constraints
6. The embedder resizes the hosted window to accept the `Scene`.